server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,581 @@
+            ".prefix"
+        ):
+            hier.add_unexpected_entries(
+                Hierarchy.NON_PREFIXES, controller, prefix_path.name
+            )
+        elif prefix_path.name.startswith("prefix."):
+            hier.add_unexpected_entries(
+                Hierarchy.WRONG_PREFIXES, controller, prefix_path.name
+            )
+
+    return 0
+
+
+# 'hier' is an object of ArchiveHierarchy class imported from hierarchy.py
+def verify_archive(hier):
+
+    controllers = Path(hier.path).glob("*")
+    verify_valid_controllers(hier, controllers)
+
+    # now check each "good" controller and get the tarballs it contains
+    for controller in hier.controllers:
+        direct_entries = glob.iglob(os.path.join(hier.path, controller, "*"))
+        hidden_entries = Path(hier.path, controller).glob(".*")
+        if hidden_entries:
+            for hid_entries_path in hidden_entries:
+                if hid_entries_path.is_file():
+                    hier.add_unexpected_entries(
+                        Hierarchy.UNEXPECTED_OBJECTS, controller, hid_entries_path.name,
+                    )
+        controller_subdir = list()
+        for item in direct_entries:
+            item_path = Path(item)
+            if item_path.is_dir():
+                controller_subdir.append(item_path.name)
+            elif item_path.is_symlink():
+                try:
+                    item_p = item_path.resolve(strict=True)
+                except FileNotFoundError:
+                    hier.add_error_or_inconceivable_entries(
+                        Hierarchy.ERROR,
+                        controller,
+                        f"{item_path.name}_path, '{item_path}', does not resolve to a real location",
+                    )
+                symlink_item = f"{item_path.name} -> {item_p}"
+                hier.add_unexpected_entries(
+                    Hierarchy.UNEXPECTED_SYMLINKS, controller, symlink_item
+                )
+            elif all(
+                [
+                    item_path.is_file(),
+                    not item_path.name.endswith(".tar.xz"),
+                    not item_path.name.endswith(".tar.xz.md5"),
+                ]
+            ):
+                hier.add_unexpected_entries(
+                    Hierarchy.UNEXPECTED_OBJECTS, controller, item_path.name
+                )
+            elif item_path.is_file() and (
+                item_path.name.endswith(".tar.xz")
+                or item_path.name.endswith(".tar.xz.md5")
+            ):
+                hier.add_tarballs(controller)
+            else:
+                hier.add_error_or_inconceivable_entries(
+                    Hierarchy.INCONCEIVABLE, controller, item_path.name,
+                )
+        verify_subdirs(hier, controller, controller_subdir)
+        verify_prefixes(hier, controller)
+
+    return 0
+
+
+# Incoming Hierarchy
+def verify_tar_dirs(ihier, tarball_dirs, tblist, controller):
```

Comment:
Do you really want to have `tblist` as a parameter?  It appears that both callers of this function provide the same value for it.  And, the name doesn't suggest that it is a _callable_, so it took me a moment to figure out what is happening when it's used....

Unless you foresee another use for it or feel the distinct need for the flexibility, I would remove it and just call `ihier.add_unexpected_entries()` directly.

Reply:
Also, `tarball_dirs` doesn't strike me as a good name for this parameter:  it's not the name or path of a directory containing tarballs, it's a _list of tarball names_.